### PR TITLE
Add create new group api handler to groups package.

### PR DESCRIPTION
**Pull request by mike-kelly-methods**
### What

As per title.

### How to review

Ensure that all builds and tests pass. Review code.

### Who can review

Anyone but me.

### CODE REVIEW

**Comment by AWitcherONS** [comment]
make test all passed
make test-component all passed
code format nothing serious just alignments
code looks logical and understandable (from a novice point of view for maintenance etc )

local postman 
1) erroneous json passed
2) happy path passed
3) duplicate check passed
4) without name passed
5) no description passed
6) no precedence   passed

**Review by AWitcherONS — DISMISSED** [review]
make test all passed
make test-component all passed
code format nothing serious just alignments
code looks logical and understandable (from a novice point of view for maintenance etc )

local postman

erroneous json passed
happy path passed
duplicate check passed
without name passed
no description passed
no precedence passed

File: api/groups.go [hunk 1]
```diff
@@ -11,6 +11,44 @@ import (
...
+		return nil, models.NewErrorResponse(http.StatusBadRequest, nil, validationErrs...)
+	}
+
+	// build create group input
+	createGroup.GenerateGroupName()
+	input := createGroup.BuildCreateGroupInput(&api.UserPoolId)
+	_, err = api.CognitoClient.CreateGroup(input)
+	if err != nil {
+		cognitoErr := models.NewCognitoError(ctx, err, "Cognito CreateGroup request from create a new group endpoint")
+		if cognitoErr.Code == models.GroupExistsError {
```
Comment (mwnicks) [inline api/groups.go:38]: Are there any other scenarios that should return BadRequest? (e.g. invalid precedence value)
Or does the internal validation cover all of these?

File: models/groups.go [hunk 2]
```diff
@@ -136,3 +143,63 @@ func (g *Group) BuildSuccessfulJsonResponse(ctx context.Context) ([]byte, error)
...
+	Precedence *int64 `json:"precedence"`
+	GroupName string
+}
+
+func (g *CreateGroup) ValidateCreateGroupRequest(ctx context.Context) []error {
+	var validationErrs []error
+
+	if g.Description == nil {
+		validationErrs = append(validationErrs, NewValidationError(ctx, InvalidGroupDescription, MissingGroupDescription))
+	} else {
```
Comment (mwnicks) [inline models/groups.go]: Could this be simplified to this

} else if m, _ := regexp.MatchString("(?i)^role_.*", *g.Description); m {

File: models/groups.go [hunk 3]
```diff
@@ -136,3 +143,63 @@ func (g *Group) BuildSuccessfulJsonResponse(ctx context.Context) ([]byte, error)
...
+		// check to ensure description doesn't start with reserved pattern `role_`
+		m, _ := regexp.MatchString("(?i)^role_.*", *g.Description)
+		if m {
+			validationErrs = append(validationErrs, NewValidationError(ctx, InvalidGroupDescription, IncorrectPatternInGroupDescription))
+		}
+	}
+	if g.Precedence == nil {
+		validationErrs = append(validationErrs, NewValidationError(ctx, InvalidGroupPrecedence, MissingGroupPrecedence))
+	} else {
```
Comment (mwnicks) [inline models/groups.go]: Could this not just be an else if?

**Review by mwnicks — COMMENTED** [review]
Couple of comments and a question

File: api/groups.go (hunk 1 above)
Comment (mike-kelly-methods) [reply api/groups.go:38]: Yep, internal validation covers all these scenarios.

Comment (mwnicks) [reply api/groups.go:38]: The scenario I thought wasn't covered was if the precedence was over the max, but just reread the docs and there is no max value.
Not sure where I got the max of 100 from.

**Review by mwnicks — APPROVED** [review]

**Pull request merged**